### PR TITLE
IP-Adapter face id

### PR DESCRIPTION
**Pull request by huchenlei**
## Installation of insightface
**FaceID** requires `insightface`, you need to install them in your A1111 venv. Check [this ComfyUI issue](https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/162) for help. 

In summary, building insightface from source requies compiling some C++ code, which in principle should be avoided at all cost in this repository. See https://github.com/Mikubill/sd-webui-controlnet/pull/1875. The workaround this PR has is downloading the pre-build wheel for the largest user group (Windows amd64 users). Other users will need build from source by running `pip install insightface`.

## Models
The following table shows the combination of checkpoint and preprocessor to use for each FaceID IPAdapter Model. Any Tensor size mismatch you may get it is likely caused by a wrong combination.

| SD v. | IPadapter | Preprocessor | LoRA |
|---|---|---|---|
| v1.5 | [FaceID](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid_sd15.bin) | ip-adapter_face_id | [FaceID LoRA](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid_sd15_lora.safetensors) |
| v1.5 | [FaceID Plus](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plus_sd15.bin) | ip-adapter_face_id_plus | [FaceID Plus LoRA](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plus_sd15_lora.safetensors) |
| v1.5 | [FaceID Plus v2](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plusv2_sd15.bin) | ip-adapter_face_id_plus | [FaceID Plus v2 LoRA](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plusv2_sd15_lora.safetensors) |
| SDXL | [FaceID](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid_sdxl.bin) | ip-adapter_face_id | [FaceID SDXL LoRA](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid_sdxl_lora.safetensors) |

It is recommended to load the corresponding LoRA when generating with a weight of 0.5 ~ 0.7.

### Step3: Generate
Face input:
image
Output:
image

## TODO
- [x] face_id support
- [x] face_id_plus support
- [x] install script for insightface lib
- [x] verify existing IP-adapters are not broken

## Next Steps
- Automatically apply corresponding LoRA
- Unload insightface model. (Is that necessary? We already run insightface on CPU)
- Documentation on IPAdapter usage
- Automatically pick preprocessor/model when one is selected

### CODE REVIEW

**Comment by t00350320** [comment]
Hi, I have tested your latest branch, but got this error while ip-adapter-faceid_sd15 loaded:

Missing key(s) in state_dict: "perceiver_resampler.proj_in.weight", "perceiver_resampler.proj_in.bias", "perceiver_resampler.proj_out.weight", "perceiver_resampler.proj_out.bias", "perceiver_resampler.norm_out.weight", "perceiver_resampler.norm_out.bias", "perceiver_resampler.layers.0.0.norm1.weight", "perceiver_resampler.layers.0.0.norm1.bias", "perceiver_resampler.layers.0.0.norm2.weight", "perceiver_resampler.layers.0.0.norm2.bias", "perceiver_resampler.layers.0.0.to_q.weight", "perceiver_resampler.layers.0.0.to_kv.weight", "perceiver_resampler.layers.0.0.to_out.weight", "perceiver_resampler.layers.0.1.0.weight", "perceiver_resampler.layers.0.1.0.bias", "perceiver_resampler.layers.0.1.1.weight", "perceiver_resampler.layers.0.1.3.weight", "perceiver_resampler.layers.1.0.norm1.weight", "perceiver_resampler.layers.1.0.norm1.bias", "perceiver_resampler.layers.1.0.norm2.weight", "perceiver_resampler.layers.1.0.norm2.bias", "perceiver_resampler.layers.1.0.to_q.weight", "perceiver_resampler.layers.1.0.to_kv.weight", "perceiver_resampler.layers.1.0.to_out.weight", "perceiver_resampler.layers.1.1.0.weight", "perceiver_resampler.layers.1.1.0.bias", "perceiver_resampler.layers.1.1.1.weight", "perceiver_resampler.layers.1.1.3.weight", "perceiver_resampler.layers.2.0.norm1.weight", "perceiver_resampler.layers.2.0.norm1.bias", "perceiver_resampler.layers.2.0.norm2.weight", "perceiver_resampler.layers.2.0.norm2.bias", "perceiver_resampler.layers.2.0.to_q.weight", "perceiver_resampler.layers.2.0.to_kv.weight", "perceiver_resampler.layers.2.0.to_out.weight", "perceiver_resampler.layers.2.1.0.weight", "perceiver_resampler.layers.2.1.0.bias", "perceiver_resampler.layers.2.1.1.weight", "perceiver_resampler.layers.2.1.3.weight", "perceiver_resampler.layers.3.0.norm1.weight", "perceiver_resampler.layers.3.0.norm1.bias", "perceiver_resampler.layers.3.0.norm2.weight", "perceiver_resampler.layers.3.0.norm2.bias", "perceiver_resampler.layers.3.0.to_q.weight", "perceiver_resampler.layers.3.0.to_kv.weight", "perceiver_resampler.layers.3.0.to_out.weight", "perceiver_resampler.layers.3.1.0.weight", "perceiver_resampler.layers.3.1.0.bias", "perceiver_resampler.layers.3.1.1.weight", "perceiver_resampler.layers.3.1.3.weight".

Hi, I have used your newest branch, but I got this error when using ip-adapter-faceid_sd15:

Missing key(s) in state_dict: "perceiver_resampler.proj_in.weight", "perceiver_resampler.proj_in.bias", "perceiver_resampler.proj_out.weight", "perceiver_resampler.proj_out.bias", "perceiver_resampler.norm_out.weight", "perceiver_resampler.norm_out.bias", "perceiver_resampler.layers.0.0.norm1.weight", "perceiver_resampler.layers.0.0.norm1.bias", "perceiver_resampler.layers.0.0.norm2.weight", "perceiver_resampler.layers.0.0.norm2.bias", "perceiver_resampler.layers.0.0.to_q.weight", "perceiver_resampler.layers.0.0.to_kv.weight", "perceiver_resampler.layers.0.0.to_out.weight", "perceiver_resampler.layers.0.1.0.weight", "perceiver_resampler.layers.0.1.0.bias", "perceiver_resampler.layers.0.1.1.weight", "perceiver_resampler.layers.0.1.3.weight", "perceiver_resampler.layers.1.0.norm1.weight", "perceiver_resampler.layers.1.0.norm1.bias", "perceiver_resampler.layers.1.0.norm2.weight", "perceiver_resampler.layers.1.0.norm2.bias", "perceiver_resampler.layers.1.0.to_q.weight", "perceiver_resampler.layers.1.0.to_kv.weight", "perceiver_resampler.layers.1.0.to_out.weight", "perceiver_resampler.layers.1.1.0.weight", "perceiver_resampler.layers.1.1.0.bias", "perceiver_resampler.layers.1.1.1.weight", "perceiver_resampler.layers.1.1.3.weight", "perceiver_resampler.layers.2.0.norm1.weight", "perceiver_resampler.layers.2.0.norm1.bias", "perceiver_resampler.layers.2.0.norm2.weight", "perceiver_resampler.layers.2.0.norm2.bias", "perceiver_resampler.layers.2.0.to_q.weight", "perceiver_resampler.layers.2.0.to_kv.weight", "perceiver_resampler.layers.2.0.to_out.weight", "perceiver_resampler.layers.2.1.0.weight", "perceiver_resampler.layers.2.1.0.bias", "perceiver_resampler.layers.2.1.1.weight", "perceiver_resampler.layers.2.1.3.weight", "perceiver_resampler.layers.3.0.norm1.weight", "perceiver_resampler.layers.3.0.norm1.bias", "perceiver_resampler.layers.3.0.norm2.weight", "perceiver_resampler.layers.3.0.norm2.bias", "perceiver_resampler.layers.3.0.to_q.weight", "perceiver_resampler.layers.3.0.to_kv.weight", "perceiver_resampler.layers.3.0.to_out.weight", "perceiver_resampler.layers.3.1.0.weight", "perceiver_resampler.layers.3.1.0.bias", "perceiver_resampler.layers.3.1.1.weight", "perceiver_resampler.layers.3.1.3.weight".
 
PTAL!!! Thank you!

**Comment by huchenlei** [comment]
It's still WIP. I will flip it to ready-to-review once everything is ready.

**Comment by huchenlei** [comment]
face_id working now. face_id_plus still WIP.

**Comment by huchenlei** [comment]
@sdbds I think the implementation is mostly done. I decide to not including the insightface lib source code here as in previous PR there was some issues with running insightface from source. This PR includes an install script that fetches the pre-build wheel for windows arm64 users, but for all other users, they need to manually install insightface package. There are already sufficient discussions here about insightface installation: https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/162

**Comment by sdbds** [comment]
> @sdbds I think the implementation is mostly done. I decide to not including the insightface lib source code here as in previous PR there was some issues with running insightface from source. This PR includes an install script that fetches the pre-build wheel for windows arm64 users, but for all other users, they need to manually install insightface package. There are already sufficient discussions here about insightface installation: [cubiq/ComfyUI_IPAdapter_plus#162](https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/162)

Thanks for the swift work! I will test it afterward. I noticed earlier when I found the onnx dependency that it seemed like the source code was more of a pain.

**Comment by vidosiki43** [comment]
Hello. Can you please tell me how to set preprocessor IP-Adapter face id? I have updated sd-webui-controlnet, but I don't have this preprocessor
295081734-93fe6c42-8dd2-4a0a-82b0-112d8dd61ce2

**Comment by huchenlei** [comment]
The PR has not been merged yet. You need to checkout the PR manually to use faceid or wait the PR to be merged.

**Comment by huchenlei** [comment]
The PR has not been merged yet. You need to checkout the PR manually to use faceid or wait the PR to be merged.

**Comment by huchenlei** [comment]
The PR has not been merged yet. You need to checkout the PR manually to use faceid or wait the PR to be merged.

File: scripts/controlmodel_ipadapter.py [hunk 1]
```diff
@@ -232,6 +346,16 @@ def get_image_embeds(self, clip_vision_output):
...
         return image_prompt_embeds, uncond_image_prompt_embeds
 
+    @torch.inference_mode()
+    def get_image_embeds_faceid_plus(self, face_embed, clip_vision_output):
+        face_embed = face_embed['image_embeds'].to(self.device)
+        from annotator.clipvision import clip_vision_h_uc
+        clip_embed = clip_vision_output['hidden_states'][-2].to(device='cpu', dtype=torch.float32)
+        return (
+            self.image_proj_model(face_embed, clip_embed),
+            self.image_proj_model(torch.zeros_like(face_embed), clip_vision_h_uc.to(clip_embed)),
```
Comment (huchenlei) [inline scripts/controlmodel_ipadapter.py:356]: I am not very sure whether we should also use `clip_vision_vith_uc` for SDXL here.

The ComfyUI implementation just does `torch.zeros_like(clip_embed)` here. Probably need to investigate later on what should be the correct input here.

File: scripts/processor.py [hunk 2]
```diff
@@ -710,6 +746,8 @@ def run_model(self, img, res=512, **kwargs):
     "ip-adapter_clip_sd15",
     "ip-adapter_clip_sdxl",
     "t2ia_style_clipvision"
+    "ip-adapter_clip_face_id",
```
Comment (huchenlei) [inline scripts/processor.py:749]: Typo 
- `ip-adapter_clip_face_id` -> `ip-adapter_face_id`
- `ip-adapter_clip_face_id_plus` -> `ip-adapter_face_id_plus`

**Comment by huchenlei** [comment]
@sdbds Gentle ping on the review.

**Comment by aMoonshine** [comment]
where to get preprocessor for faceID? or how to make it preprocessor, not a model if it's just another model there?
image

**Comment by jeguepowerr** [comment]
I already did git pull but the preoprocessor still doesn't appear 😔
 What am I doing wrong?

**Comment by identify86** [comment]
屏幕截图 2024-01-11 090049
how to fix it,THX!!!

**Comment by imfog** [comment]
> Hello. Can you please tell me how to set preprocessor IP-Adapter face id? I have updated sd-webui-controlnet, but I don't have this preprocessor 295081734-93fe6c42-8dd2-4a0a-82b0-112d8dd61ce2 

need **insightFace**
[https://github.com/deepinsight/insightface](url)

**Comment by identify86** [comment]
please add FaceId portrait to webui A1111,THX!!!!!!!!!!!!!!!!!!!!!!

**Comment by huchenlei** [comment]
> please add FaceId portrait to webui A1111,THX!!!!!!!!!!!!!!!!!!!!!!

it's already added several days ago: https://github.com/Mikubill/sd-webui-controlnet/discussions/2530

**Comment by identify86** [comment]
Great job,Thx!!!




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Chenlei ***@***.***&gt;; 
发送时间: 2024年1月24日(星期三) 晚上9:51
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Mikubill/sd-webui-controlnet] IP-Adapter face id (PR #2434)





  
please add FaceId portrait to webui A1111,THX!!!!!!!!!!!!!!!!!!!!!!
  
it's already added several days ago: #2530

**Pull request merged**